(design_proposal): chef cli catalog (#877)
chef/chef-workstation: docs/Chef-CLI-Catalog.md

$ chef-policy clean cookbooks
$ chef-policy group delete (and more? maybe show a policy group)
```
These sub-commands map to the previous top level policy commands. As an example, the
previous command `chef clean-policy-revisions` would map to the new command
`chef-policy clean revisions`.

Once the new `chef-policy` exists, we need to update the `chef` catalog to point to
the new binary and add a section called "Relocated Commands".

The new help command of the `chef` catalog should look like:
```
$ chef
Usage:
    chef -h/--help
    chef -v/--version
    chef command [arguments...] [options...]

Available Commands:
    env                     Prints environment variables used by Chef Workstation
    gem                     Runs the 'gem' command in context of the embedded Ruby
    exec                    Runs the command in context of the embedded ruby
    shell-init              Initialize your shell to use Chef Workstation as your primary Ruby
    generate                Generate a new repository, cookbook, or other component
    describe-cookbook       Prints cookbook checksum information used for cookbook identifier
    policy                  Manage Chef Policyfiles

> Relocated Commands:

Reviewer: ```suggestion
Renamed Commands:
```

    install                 Use now: chef policy install
    update                  Use now: chef policy update
    push                    Use now: chef policy push
    push-archive            Use now: chef policy push
    show-policy             Use now: chef policy show
    diff                    Use now: chef policy diff
    export                  Use now: chef policy export
    clean-policy-revisions  Use now: chef policy clean revisions
    clean-policy-cookbooks  Use now: chef policy clean cookbooks
    delete-policy-group     Use now: chef policy group delete
    delete-policy           Use now: chef policy delete
    undelete                Use now: chef policy undelete

Experimental Commands:
    analyze                 Analyze artifacts from a Chef Infra Server
```

#### Open questions for further discussion with UX:
* How do we want to present these sections? What is the right order?

> * What verbiage should we use to express the relotation of commands? (`Use now` Vs `Moved to`)

Reviewer: `now` is unnecessary, can just say `Use`

    apply                   Executable program that runs a single recipe
    shell                   A recipe debugging tool that allows the use of breakpoints within recipes
    berks                   Dependency manager for Chef cookbooks
    cookstyle               Code linting tool that helps you write better Chef Infra cookbooks

Relocated Commands:
    install                 Use now: chef policy install
    update                  Use now: chef policy update
    push                    Use now: chef policy push
    push-archive            Use now: chef policy push
    show-policy             Use now: chef policy show
    diff                    Use now: chef policy diff
    export                  Use now: chef policy export
    clean-policy-revisions  Use now: chef policy clean revisions
    clean-policy-cookbooks  Use now: chef policy clean cookbooks
    delete-policy-group     Use now: chef policy group delete
    delete-policy           Use now: chef policy delete
    undelete                Use now: chef policy undelete

Experimental Commands:
    analyze                 Analyze artifacts from a Chef Infra Server
```

### EXTRA: Present correct usage of sub-binaries

> An additional task from this proposal is the use of an environment variable that

Reviewer: Talked with Salim about this - IMO, our first implementation should reference `chef policy` in all the help text. This makes it more confusing for our effortless developers who will directly invoke the `chef-policy` binary (without all the overhead of the entire Chef Workstation ecosystem). But those are internal users we can help.

In the future I could see us developing a Golang `effortless` executable that functions similarly to the `chef` golang executable. It allows us to start creating a custom workflow for them, leveraging tools under the covers that they do not care whether are written in Ruby or Golang. At that time I would introduce some kind of environment variable that would allow the help text to display `chef policy` or `effortless policy`.

can control the "Usage" message of sub-binaries. Currently, the `chef-analyze`
binary displays the usage message `chef-analyze [command]` when it is run directly
and also when it is run from the `chef` catalog. We would like the usage message
to be consistent with the way users execute the commands so if you run `chef analyze`
the usage sould be `chef analyze [command]`; and if you run `chef-analyze` the
usage should continue to be `chef-analyze [command]`.

## Open points for further discussion:
* Are we discouraging the use of individual sub-binaries like `knife` or `chef-run`?

> * What happens when a user runs a sub-binary without the prefix `chef`?

Reviewer: ```suggestion
* What happens when a user runs a sub-binary without the prefix `chef`?
    * How do we capture telemetry on usage when users directly access tools? (EG, `berks`, `inspec` instead of `chef berks`, `chef inspec`)
```

previous command `chef clean-policy-revisions` would map to the new command
`chef-policy clean revisions`.

Once the new `chef-policy` exists, we need to update the `chef` catalog to point to
the new binary and add a section called "Relocated Commands".

The new help command of the `chef` catalog should look like:
```
$ chef
Usage:
    chef -h/--help
    chef -v/--version
    chef command [arguments...] [options...]

Available Commands:
    env                     Prints environment variables used by Chef Workstation
    gem                     Runs the 'gem' command in context of the embedded Ruby
    exec                    Runs the command in context of the embedded ruby
    shell-init              Initialize your shell to use Chef Workstation as your primary Ruby
    generate                Generate a new repository, cookbook, or other component
    describe-cookbook       Prints cookbook checksum information used for cookbook identifier
    policy                  Manage Chef Policyfiles

Relocated Commands:
    install                 Use now: chef policy install
    update                  Use now: chef policy update
    push                    Use now: chef policy push

>     push-archive            Use now: chef policy push-archive

Reviewer: Going with the CLI UX `noun verb` guidelines I think the following make more sense:

```
chef policy install
chef policy update
chef policy push
chef policy show
chef policy diff
chef policy export
chef policy delete
chef policy undelete
chef policy-archive push
chef policy-cookbooks clean
chef policy-group delete
chef policy-revisions clean
```

Under the covers, `chef policy-group delete` could still invoke `chef-policy group delete` but that would be an internal implementation detail. It would make it easier for us to implement because it would be less changing of existing code. But I think the public contract is more important to cover in this design doc than the internal implementation of our command routing.

Thoughts?

Author: From our conversation over zoom we have agreed that these commands look correct with the exception of `chef-policy push-archive` that will be removed and include into the `chef-policy push` command.

Reviewer: More context - I convinced myself that switching to a strictly enforced 'noun verb' syntax means we end up with a bunch of unnecessary nouns (`policy-revision`, `policy-cookbooks`, `policy-groups`, etc.) with 1 verb each.

Instead we talked over some of the existing commands and think the PR correctly documents them. The 'noun' in this case is always 'policy' because all these commands are all connected to Policyfiles.

The only tricky one is `chef-policy clean cookbooks` because it deletes cookbooks that are not referenced by a policyfile. Maybe this becomes an alias for something like `chef cookbooks clean`?

    export                  Use now: chef policy export
    clean-policy-revisions  Use now: chef policy clean revisions
    clean-policy-cookbooks  Use now: chef policy clean cookbooks
    delete-policy-group     Use now: chef policy group delete
    delete-policy           Use now: chef policy delete
    undelete                Use now: chef policy undelete

Experimental Commands:
    analyze                 Analyze artifacts from a Chef Infra Server
```

### EXTRA: Present correct usage of sub-binaries

An additional task from this proposal is the use of an environment variable that
can control the "Usage" message of sub-binaries. Currently, the `chef-analyze`
binary displays the usage message `chef-analyze [command]` when it is run directly
and also when it is run from the `chef` catalog. We would like the usage message
to be consistent with the way users execute the commands so if you run `chef analyze`
the usage sould be `chef analyze [command]`; and if you run `chef-analyze` the
usage should continue to be `chef-analyze [command]`.

## Open points for further discussion:
* Are we discouraging the use of individual sub-binaries like `knife` or `chef-run`?

> * What happens when a user runs a sub-binary without the prefix `chef`?

Reviewer: IMO, we should not communicate the existence of these binaries to users but we should not prevent their usage. It is just unsupported usage.

Reviewer: I misunderstood this line - I was thinking that we should document `chef policy` and not document `chef-policy` but if they run `chef-policy` we don't do anything to prevent them from doing that.

# Chef CLI (Catalog)

Throughout the history of Chef, we have created and acquired a variety

> of tools that have grown to a point that, as a Chef practitioner, it is

Reviewer 2: Chef Operator to bring this in-line with our current target user base.

hard to know, discover and understand all of them. A few of the tools

> that our users use (almost) every day are:

Reviewer 3: I think having a catalog of all available commands/tools is very important. Once the user becomes aware of a tool, the next step is really for them to figure out what it does and if it's useful to them. Is that part of the intended scope of this work? If so then we can collaborate w/ UX team on figuring where/how to display relevant information about each tool and maybe include URLs to more detailed information (e.g. docs page) to users

Author: Yes, this is exactly one of the ideas/goals and therefore, collaboration with UX will be required.

* chef (CLI that includes Policyfiles management)
* hab (Habitat CLI)
* inspec
* ohai
* kitchen (Test-kitchen)
* knife
* berks (Berkshelf)
* cookstyle
* delivery
* chef-apply
* chef-client
* chef-run
* chef-shell
* chef-solo

## Goals
* Have a single, unified way to discover tools to interact with Chef products

> * Huge speed improvements

Reviewer 2: All of the other goals are more tangible. I'd say something like: "Consistent/Usable performance on all supported platforms"

* Gather data about how our users consume and interact with our tools (Telemetry)
* Standardize the UX of our (CLI) tools
* Have the flexibility to rename tools and commands within our ecosystem
* Enable developers to use any supported programming languages at Chef for efficiency

## Motivation

    As a Chef practitioner,
    I would like to have a single Command Line Interpreter (CLI) that lets me easily interact with every Chef product.
    I would also like to have a way to discover new tools that can help me with my day to day work.
    Finally, since I run these tools daily, I need them to run fast

    As a Chef developer,

>     I would have a tremendous benefit from this architecture since

Reviewer 2: These I woulds are future tense assuming this has been implemented. I think they need to be written as what is desired. It's also pre-judging the architecture before garnering feedback. I'd suggest more along the lines of: I would like a more flexible architecture that allows me to invest in improvements selectively where they will have the most impact.

    I would be able to gather information about how our users consume and interact with our tools,
    as well as identifying the commands that are used the most,

>     so I can improve the speed and/or do extreme modifications to the code base (do a rewrite in a different, more efficient, programing language)

Reviewer 2: This so states things you could do not why you'd want to. Is there a way to say this without talking about extreme changes. Anytime we use that kind of language it can alienate folks that want more measured changed/no change. It's ok if you really mean to drive extreme change and understand the above, but make sure that's the intent.

## Specification

> Currently, inside the latest version of Chef Workstation (0.14.16 as of 01/20/2020),

Reviewer 2: It's been there longer, so I'd find when it landed or just remove the version.

Author: yup, since `0.10.41`

> the team has implemented a top-level chef command that acts as a wrapper around our

Reviewer 2: s/chef/`chef`

> variety of tools. The main idea about this command is to have the ability to create

Reviewer 2: wrapper around chef commands.

Reviewer 3: is it correct to assume the wrapper has been applied to all commands for all binaries included in workstation?

> variety of tools. The main idea about this command is to have the ability to create

Reviewer 2: Here you are future tense again as if it's been done. I think you want to say something like this design proposes we expand this scope to be a proxy/catalog of all the user facing cli tools we ship.

a catalog of tools/sub-commands, it will act as a proxy between the user and our
tools.

Quick diagram that illustrates the new Chef CLI Catalog:

![chef-cli-catalog](img/chef-top-level-command.jpg)

> With this architecture, we are now able to integrate and modularize multiple tools

Reviewer 2: s/we are now able/we would be able

in a single place and make the chef command a first citizen of our development
experience. We will also be able to gather immediate information from our users of
all our tools so we can understand what parts of our tooling are the most used and
which areas can we improve.

Once detecting such areas of improvement, we can have the flexibility to improve
tools entirely or sub-sections of the tool. Say for instance, that our users run
`knife search` around 30-50 times a day, if the command takes around 10-15

> seconds to run, we are consuming ~10 minutes from our users only on waiting time.

Reviewer 2: ~10 minutes a day from our users waiting for value.

Reviewer 3: Can someone help me understand how catalog improvements are links to performance? (i.e. how does the catalog work speeds things up)

With this approach we could rewrite that specific command to run 10x faster and
give that time back to our users.

![mocked-chef-help-command](img/mocked-chef-help-command.png)

[Gist with a few mocked help commands.](https://gist.github.com/afiune/1dc854089002e182288a0452eaa91908)

## Downstream Impact
TBA

## Milestones

> ### Implement Telemetry into the top-level chef command

Reviewer 3: Just FYI, there's some dependency here on business requirements and legal requirements. Essentially we need to solidify our business requirements, get buy in from potential stakeholders (i.e. outside of Prod/Eng) and then get legal approval to proceed. This work is essential but just want to call out that the requirements are a little TBD right now and we have legal hurdles to overcome.

We need to start gathering information about how our users use the chef CLI,
this will allow us to understand the impact that we will have from the

> aggressivemoves we are about to do.

Reviewer 2: Can we do this in a non-aggressive way?

Reviewer 3: With regard to telemetry, how can we link this to legal requirements around data capture so the design doc includes the our internal privacy requirements? (e.g. how can we log requirements from legal such as only turn on anonymous data capture after accepting the EULA?)

    analyze                 Analyze artifacts from a Chef Infra Server
```

### Automation of reference documentation

Motivation:

    As a technical writer
    I want to add examples to the chef catalog documentation

>     so that users will know the tricky bits and we the documentations gets updated automatically as the code evolves

Author: ```suggestion
    so that users will know the tricky bits and the documentation gets updated automatically as the code evolves
```
